design-proposals/vm-delete-protection.md

```diff
@@ -0,0 +1,155 @@
+
+## Goals
+
+- Establish a delete protection mechanism for VirtualMachine objects.
+
+## Non Goals
+
+- Create a delete protection for VirtualMachineInstance objects.
+- Create a delete protection for virt-launcher Pod objects.
+
+## Definition of Users
+
+Everyone who has the permissions to delete VMs.
+
+## User Stories
+
+As a user who has the permissions to delete VMs, I want to protect my VMs from accidental deletion.
+
+## Repos
+
+kubevirt/kubevirt
+
+# Design
+
+The design is divided into two sections. The first section shows how the VM protection can be enabled by a user, the
+second section depicts possible implementations of the feature in the backend.
+
+### Enable the VM Object Protection
+
+To protect VMs against deletion, they need to be annotated with  `kubevirt.io/vm-delete-protection` set to `true` or
+`True`. If the annotation is present with one of the aforementioned values, requests to delete this VM will be rejected
+by KubeVirt. By default, this annotation is not present on a VM. It is up to the user to enable the deletion protection.
+The annotation looks like this in the VM manifests:
+
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  annotations:
+    kubevirt.io/vm-delete-protection: "true"
+spec:
+[...]
+```
+
+### Backend
+
+#### Option \#1: ValidatingAdmissionPolicy
+
+The VM protection is implemented by adding
+a  [ValidationAdmissionPolicy and ValidationAdmissionPolicyBinding](https://kubernetes.io/docs/reference/access-authn-authz/validating-admission-policy/)
+objects. On every delete action against a VM object, the ValidatingAdmissionPolicy will check for presence of the
+annotation. If it is present and its value equals “true” or “True” the request will be rejected. The
+ValidationAdmissionPolicy and ValidationAdmissionPolicyBinding objects will be configured with the following parameters:
+
+- matchConstrains: Apply to \`virtualmachine\` objects; Operations: DELETE.
+- Variables:
+    - \`annotation\`: Returns the VM annotation field;
+        - string('kubevirt.io/vm-delete-protection')
+- Validation rules:
+    - Checks if the annotation is present at all, if present, checks if it is equal to “true” or “True”;
+        - (\!(variables.annotation in oldObject.metadata.annotations) || \!oldObject.metadata.annotations\[variables.annotation\].matches('^(true|True)$'))
+
+Check [https://github.com/kubevirt/community/blob/main/design-proposals/shadow-node.md\#backwards-compatibility](https://github.com/kubevirt/community/blob/main/design-proposals/shadow-node.md#backwards-compatibility)
+for backwards compatibility details.
+
+Pros:
+
+- Kubernetes built-in feature which is straightforward to maintain, implement and configure.
+- ValidationAdmisssionPolicy is [GA](https://kubernetes.io/blog/2024/03/12/kubernetes-1-30-upcoming-changes/) since Kuberentes version 1.30.
+
+Cons:
+
+- It will not work in deployments using Kubernetes versions \< 1.30
```

Comment:
I'd not consider this as a con because this is not a requirement to work in older Kubernetes versions.

Reply:
You are right, moreover, in the best case this feature will be onboarded in version 1.5, and according to https://github.com/kubevirt/sig-release/blob/main/releases/k8s-support-matrix.md, unless exception, we will support K8S ranging from 1.30 to 1.32.
Dropped.